File: compiler/src/cobalt/AST/IR.hs
module AST.IR where

import Data.Scientific

import qualified JVM.ClassFile

data ModuleIR = ModuleIR ModuleHeaderIR [ModelIR]
    deriving (Show)

data ModuleHeaderIR = ModuleHeaderIR
    { modNameSpace :: NameSpaceIR
    , modImports :: [ImportIR]
    }
    deriving (Show, Eq)

data MethodIR = MethodIR
    { methodName :: NameIR
    , methodAnns :: [AnnotationIR]
    , methodParams :: [FieldIR]
    , methodModifiers :: [ModifierIR]
    , methodReturnType :: (Maybe TypeIR)
    , methodBody :: BlockIR
    }
    deriving (Show, Eq)

data ConstantIR = ConstantIR
    deriving (Show, Eq)

data ModelTypeIR
    = ClassModelIR
    | ObjectModelIR
    | TraitModelIR
    | UnknownModelIR
    deriving (Eq, Show)

data ModifierIR
    = PublicIR
    | ProtectedIR
    | PrivateIR
    | PackageLocalIR
    | AbstractIR
    | FinalIR
    | PureIR
    deriving (Eq, Show)

data ModelIR = ModelIR
    { modelName :: NameIR
    , modelType :: ModelTypeIR
    , modelModifiers :: [ModifierIR]
    , modelFields :: [FieldIR]
    , modelParent :: Maybe TypeIR
    , modelParentArguments :: [StmtIR]
    , modelInterfaces :: [TypeIR]
    , modelBody :: StmtIR
    }
    deriving (Show, Eq)

data FieldIR = FieldIR
    { fieldName :: NameIR
    , fieldType :: Maybe TypeIR
    , fieldInit :: Maybe ExprIR
    }
    deriving (Show, Eq)

data TypeIR
    = InitIR
    | TypeRefIR RefIR
    | TypeAppIR RefIR [TypeIR] -- type application, aka Map<A,B> -> `TyApp (RefLocal "Map") [TyRef (RefLocal "A"), TyRef (RefLocal "B")]`
    | TypeRelIR TypeRelIR TypeIR TypeIR -- this allows things like <T extends Something> which would be `TyRel Extends (TyRef (RefLocal "T")) (TyRef (RefLocal "Something"))`
    deriving (Show, Eq)

data RefIR
    = RefSpecialIR SpecialRefIR
    | RefLocalIR NameIR
    | RefQualIR QualNameIR
    deriving (Show, Eq)

data SpecialRefIR
    = SuperIR
    | ThisIR
    deriving (Show, Eq)

data TypeRelIR
    = InheritsIR
    | ExtendsIR
    | EqualsIR
    deriving (Show, Eq)

data NameSpaceIR = NameSpaceIR [String]
    deriving (Show, Eq)

data NameIR = NameIR String
    deriving (Show, Eq)

data ImportIR = ImportIR [String]
    deriving (Show, Eq)

data AnnotationIR = AnnotationIR NameIR
    deriving (Show, Eq)

data QualNameIR = QualNameIR NameSpaceIR NameIR
    deriving (Show, Eq)

data IntConstantIR = IntConstantIR Integer

data BlockIR
    = InlineIR ExprIR
    | DoBlockIR StmtIR
    deriving (Show, Eq)

data ExprIR
    = BlockExprIR [ExprIR]
    | IdentifierIR NameIR
    | MethodCallIR NameIR ExprIR
    | NewClassInstanceIR TypeIR ExprIR (Maybe StmtIR)
    | StringLiteralIR String
    | TernaryIR ExprIR ExprIR ExprIR
    | TupleIR ExprIR
    | BoolConstIR Bool
    | NotIR ExprIR
    | BBinaryIR BBinOpIR ExprIR ExprIR
    | RBinaryIR RBinOpIR ExprIR ExprIR
    | IntConstIR Integer
    | DoubleConstIR Scientific
    | FloatConstIR Double
    | LongConstIR Scientific
    | NegIR ExprIR
    | ABinaryIR ABinOpIR ExprIR ExprIR
    | ArrayIR ArrayOpIR ExprIR ExprIR
    | SpecialRefAsExprIR SpecialRefIR
    deriving (Show, Eq)

data StmtIR
    = ForIR ExprIR ExprIR ExprIR StmtIR
    | WhileIR ExprIR StmtIR
    | IfIR ExprIR StmtIR (Maybe StmtIR)
    | AssignIR NameIR (Maybe TypeIR) Bool BlockIR
    | AssignMultipleIR [NameIR] (Maybe TypeIR) Bool BlockIR
    | ReassignIR NameIR BlockIR
    | ReturnIR StmtIR
    | LambdaIR [FieldIR] BlockIR
    | ModelDefIR ModelIR
    | MethodDefIR MethodIR
    | ExprAsStmtIR ExprIR
    | BlockStmtIR [StmtIR]
    | MatchIR ExprIR [CaseIR]
    | PrintIR ExprIR JVM.ClassFile.FieldType
    | PrintlnIR ExprIR JVM.ClassFile.FieldType
    deriving (Show, Eq)

data CaseIR
    = CaseIR ExprIR BlockIR
    deriving (Show, Eq)

data BBinOpIR
    = AndIR
    | OrIR
    deriving (Show, Eq)


data ArrayOpIR
    = ArrayAppendIR
    deriving (Show, Eq)

data RBinOpIR
    = GreaterEqualIR
    | GreaterIR
    | LessEqualIR
    | LessIR
    deriving (Show, Eq)

data ABinOpIR
    = AddIR JVM.ClassFile.FieldType
    | SubtractIR JVM.ClassFile.FieldType
    | MultiplyIR JVM.ClassFile.FieldType
    | DivideIR JVM.ClassFile.FieldType
    deriving (Show, Eq)

File: src/main/scala/cobalt/ir/IR.scala
package cobalt.ir

object IR {

  case class ModuleIR(header: ModuleHeaderIR, models: Seq[StatementIR])

  case class ModuleHeaderIR(nameSpace: NameSpaceIR, imports: Seq[ImportIR])

  case class ImportIR(loc: Seq[NameIR])

  case class FieldIR(name: NameIR, `type`: TypeIR, init: Option[TypeIR])

  trait TypeIR
  case class InitIR() extends TypeIR
  case class TypeRefIR(ref: RefIR) extends TypeIR
  case class TypeAppIR(ref: RefIR, types: Seq[TypeIR]) extends TypeIR

  trait RefIR
  case class RefSpecialIR(specialRef: SpecialRefIR) extends RefIR
  case class RefLocalIR(name: NameIR) extends RefIR
  case class RefQualIR(qualName: QualNameIR) extends RefIR

  trait SpecialRefIR
  case class SuperIR() extends SpecialRefIR
  case class ThisIR() extends SpecialRefIR

  trait TypeRelIR
  case class InheritsIR() extends TypeRelIR
  case class ExtendsIR() extends TypeRelIR
  case class EqualsIR() extends TypeRelIR

  case class NameSpaceIR(nameSpace: Seq[NameIR])

  case class NameIR(value: String)
  case class QualNameIR(nameSpace: NameSpaceIR, name: NameIR)

  case class AnnotationIR(name: NameIR)

  trait ModifierIR
  case object PublicIR extends ModifierIR
  case object ProtectedIR extends ModifierIR
  case object PrivateIR extends ModifierIR
  case object PackageLocalIR extends ModifierIR
  case object AbstractIR extends ModifierIR
  case object FinalIR extends ModifierIR
  case object PureIR extends ModifierIR

  // TODO Update args
  trait ExpressionIR
  case class BlockExprIR(expressions: Seq[ExpressionIR]) extends ExpressionIR
  case class IdentifierIR(name: NameIR) extends ExpressionIR
  case class MethodCallIR(name: NameIR, expression: ExpressionIR) extends ExpressionIR
  case class NewClassInstanceIR(`type`: TypeIR, expression: ExpressionIR, anonymousClass: Option[StatementIR]) extends ExpressionIR
  case class StringLiteralIR(value: String) extends ExpressionIR
  case class TernaryIR() extends ExpressionIR
  case class TupleIR() extends ExpressionIR
  case class BoolConstIR(value: Boolean) extends ExpressionIR
  case class NotIR() extends ExpressionIR
  case class ABinaryIR(op: ABinOpIR, expression1: ExpressionIR, expression2: ExpressionIR) extends ExpressionIR
  case class BBinaryIR(op: BBinOpIR, expression1: ExpressionIR, expression2: ExpressionIR) extends ExpressionIR
  case class RBinaryIR(op: RBinOpIR, expression1: ExpressionIR, expression2: ExpressionIR) extends ExpressionIR
  case class IntConstIR(value: Int) extends ExpressionIR
  case class DoubleConstIR(value: Double) extends ExpressionIR
  case class FloatConstIR(value: Float) extends ExpressionIR
  case class LongConstIR(value: Long) extends ExpressionIR
  case class NegIR(expression: ExpressionIR) extends ExpressionIR

  case class ArrayIR() extends ExpressionIR
  case class SpecialRefAsExprIR() extends ExpressionIR

  trait BlockIR extends StatementIR
  case class InlineIR(expression: ExpressionIR) extends BlockIR
  case class DoBlockIR(statement: StatementIR) extends BlockIR

  trait StatementIR
  case class ClassModelIR(name: NameIR, modifiers: Seq[ModifierIR], fields: Seq[FieldIR], parent: Option[TypeIR], parentArguments: Seq[ExpressionIR], interfaces: Seq[TypeIR], body: StatementIR) extends StatementIR
  case class ObjectModelIR(name: NameIR, modifiers: Seq[ModifierIR], fields: Seq[FieldIR], parent: Option[TypeIR], parentArguments: Seq[ExpressionIR], interfaces: Seq[TypeIR], body: StatementIR) extends StatementIR
  case class TraitModelIR(name: NameIR, modifiers: Seq[ModifierIR], fields: Seq[FieldIR], parent: Option[TypeIR], parentArguments: Seq[ExpressionIR], interfaces: Seq[TypeIR], body: StatementIR) extends StatementIR
  case class MethodIR(name: NameIR, annotations: Seq[AnnotationIR], fields: Seq[FieldIR], modifiers: Seq[ModifierIR], returnType: Option[TypeIR], body: BlockIR) extends StatementIR
  case class ForIR() extends StatementIR
  case class WhileIR() extends StatementIR
  case class IfIR(condition: ExpressionIR, ifBlock: StatementIR, elseBlock: Option[StatementIR]) extends StatementIR
  case class AssignIR(name: NameIR, `type`: Option[TypeIR], immutable: Boolean, block: BlockIR) extends StatementIR
  case class AssignMultipleIR(name: Seq[NameIR], `type`: Option[TypeIR], immutable: Boolean, block: BlockIR) extends StatementIR
  case class ReassignIR(name: NameIR, block: BlockIR) extends StatementIR
  case class ReturnIR() extends StatementIR
  case class LambdaIR() extends StatementIR
  case class ModelDefIR() extends StatementIR
  case class ExprAsStmtIR(expression: ExpressionIR) extends StatementIR
  case class BlockStmtIR(statements: Seq[StatementIR]) extends StatementIR
  case class MatchIR() extends StatementIR
  case class PrintIR() extends StatementIR
  case class PrintlnIR() extends StatementIR

  case class CaseIR(expression: ExpressionIR, block: BlockIR)

  trait OperatorIR

  trait ABinOpIR extends OperatorIR
  case object AddIR extends ABinOpIR
  case object SubtractIR extends ABinOpIR
  case object MultiplyIR extends ABinOpIR
  case object DivideIR extends ABinOpIR

  trait BBinOpIR extends OperatorIR
  case object AndIR extends BBinOpIR
  case object OrIR extends BBinOpIR

  trait RBinOpIR extends OperatorIR
  case object GreaterEqualIR extends RBinOpIR
  case object GreaterIR extends RBinOpIR
  case object LessEqualIR extends RBinOpIR
  case object LessIR extends RBinOpIR
  case object EqualIR extends RBinOpIR

}
